### PR TITLE
Make modes rememberable.

### PR DESCRIPTION
**Pull request by aartaka**
This adds `rememberable-p` slot to `root-mode` and (all) the inheriting
modes. It's set to nil for those modes that we don't want to be
remembered by `auto-mode`, like `internal-buffer`-styling modes or
one-shot modes, like `visual-mode` or `expedition-mode`.

# Things To Discuss
This, however, shifts some of the responsibilities of `auto-mode` back
to modes, instead of universal `mode-invocation`s. Should it be that
way? I'm not sure.

# How Has This Been Tested
- Nyxt compiles and runs with this. Haven't tested all the non-rememberable modes, though.

How does that look?

### CODE REVIEW

File: source/auto-mode.lisp [hunk 1]
```diff
@@ -51,16 +51,12 @@ Package prefix is optional.")
...
-(defun rememberable-of (modes auto-mode)
-  "Filter MODES based on rememberability by AUTO-MODE."
-  (if auto-mode
-      (set-difference (mode-invocations modes)
-                      (non-rememberable-modes auto-mode)
-                      :test #'equals)
-      modes))
+(defun rememberable-of (modes)
+  "Filter MODES based on `rememberable-p', turn them into `mode-invocation's."
+  (mode-invocations (remove-if-not #'rememberable-p modes)))
```
Comment (Ambrevar) [inline source/auto-mode.lisp:59]: To be picky, `remove-if-not` is deprecated.
You can use `remove-if (complement ...)` or `sera:filter` instead.

File: source/auto-mode.lisp [hunk 2]
```diff
@@ -310,11 +295,11 @@ For the storage format see the comment in the head of your `auto-mode-rules-data
       (setf url (url url)))
     (add-modes-to-auto-mode-rules
      (url-infer-match url)
-     :include (set-difference (mode-invocations (modes (current-buffer)))
+     :include (set-difference (rememberable-of (modes (current-buffer)))
```
Comment (Ambrevar) [inline source/auto-mode.lisp:298]: I don't underestand why this `mode-invocations` is replaced by `rememberable-of`.
Same below.
Can you elaborate?

**Comment by Ambrevar** [comment]
Looks good to me!
I think it's actually a very welcome change.

About moving the feature from auto-mode to mode, I think it's actually better this way, because "rememberable-p" is not a name that necessarily refers to auto-mode and can potentially be used by other modes in the future.  In other words, it's a slot that composes well.
Should we ever use rememberable-p for something other than auto-mode, all we would have to do is updat the documentation ;)

File: source/auto-mode.lisp (hunk 2 above)
Comment (aartaka) [reply source/auto-mode.lisp:298]: Sure I can. 
1. We cannot check rememberability of the modes in `add-modes-to-auto-mode-rules`, as we cannot call `rememberable-p` on `mode-instance`s passed to `add-modes-to-auto-mode-rules` (actually, we can, but that requires calling/instantiating the mode, and that's too much of state change). Because of that, we need to filter modes by rememberability _before_ they are passed to `add-modes-to-auto-mode-rules`.
2. We need to pass `mode-instance` list to `add-modes-to-auto-mode-rules`, and I've made `rememberable-of` to call `mode-instances` implicitly. Is that the right choice? I'm not sure.

Both of these can easily be solved if we check rememberability in `add-modes-to-auto-mode-rules`. Let me test how the mode instantiation works for that and I'll push it.

File: source/auto-mode.lisp (hunk 1 above)
Comment (aartaka) [reply source/auto-mode.lisp:59]: Didn't know about that 0_o Will fix!

**Comment by aartaka** [comment]
Okay, instantiating modes works for `rememberable-p` check. All fixed now :)

**Comment by aartaka** [comment]
Should this be merged now or after 2.0? `root-mode` slots changed, after all -- and it's an API change.

**Comment by Ambrevar** [comment]
I'd say now.

File: source/auto-mode.lisp (hunk 2 above)
Comment (Ambrevar) [reply source/auto-mode.lisp]: When you wrote `mode-instance` I suppose you meant `mode-invocation`s?

Anyways, the new patch introduces less changes and is clearer, so I guess we are all good!

**Comment by Ambrevar** [comment]
Wait, actually I just realized that we may be instantiating modes for nothing here, which is probably not desirable.

First thing, the easy part: in enable-matching-modes you call `rememberable-of` over a list of modes, so no need to wrap with `mode-invocation`, right?

Then, in `add-modes-to-auto-mode-rules` it depends: sometimes we have the modes, sometimes just the invocations.
What shall we do when we only have the invocations?  What about getting the default value of the mode slot with moptilities?

**Comment by aartaka** [comment]
> Wait, actually I just realized that we may be instantiating modes for nothing here, which is probably not desirable.
> 
> First thing, the easy part: in enable-matching-modes you call `rememberable-of` over a list of modes, so no need to wrap with `mode-invocation`, right?

Hmmm... Yes, good point. How about this: make `rememebrable-of` a simple `remove-if` using `rememberable-p` as filtering function, while defining additions `rememberable-p` over `mode-invocation`?

> Then, in `add-modes-to-auto-mode-rules` it depends: sometimes we have the modes, sometimes just the invocations.
> What shall we do when we only have the invocations? What about getting the default value of the mode slot with moptilities?

1. We always have `mode-invocation`s in `add-modes-to-auto-mode-rules`, according to its type declaration.
2. In case of invocations, they can also have arguments changing mode behavior, so we _need_ to instantiate them to know for sure whether they are rememberable or not.
3. Moptilities can help us there, but they don't concider the change that additional arguments bring (see 2). It's the same as with buffer instantiation that we do to know the exact value of the buffer slot at runtime.

To sum up, what I'll do:
- Add `rememberable-p` method over `mode-invocation` and simplify `rememebrable-of`.
  - Use arguments to mode instantiation in `rememberable-p`.

**Comment by aartaka** [comment]
See fae5c2c1db01c95bc59802109e3cba0a5458da96.
